### PR TITLE
Adding PPPM and Ewald solvers for electric dipoles and magnetic spins

### PR DESCRIPTION
**Pull request by julient31**
**Summary**

The PR adds two solvers, PPPM and Ewald sums, for long-range dipole-dipole interactions. Those solvers apply to electric dipoles and magnetic spins.
In addition, the corresponding short-range components were added in the SPIN package (pair_spin_dipole_cut.cpp/h and pair_spin_dipole_long.cpp/h, similar to dipole/cut and dipole/long for electric charges). 
The documentations was updated to account for those changes in:
- doc/src/pair_spin_dipole.txt
- doc/src/kspace_style.txt

Examples were provided in: examples/SPIN/dipole_spin/

**Related Issues**

This PR is an improvement of the KSPACE and SPIN packages.

**Author(s)**

Stan Moore (stamoor@sandia.gov)
Julien Tranchida (jtranch@sandia.gov)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/R

**Implementation Notes**

The results were checked by comparison to the ewald/disp method for dipoles. 
Very good agreement was found for all quantities, except the per-atom virial components obtained from the P3M solver (although the total pressure is in very good agreement). This point will later receive further investigation.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by stanmoore1** [comment]
@akohlmey @julient31 I believe this is ready to merge.

**Comment by akohlmey** [comment]
> @akohlmey @julient31 I believe this is ready to merge.

@stanmoore1 better to wait until after the upcoming stable release, though.

File: src/SPIN/fix_nve_spin.cpp [hunk 1]
```diff
@@ -296,6 +305,13 @@ void FixNVESpin::initial_integrate(int /*vflag*/)
     }
   }
 
+  // update fm_kspace if long-range
+  // remove short-range comp. of fm_kspace
+
+  if (long_spin_flag) {
+
```
Comment (akohlmey) [inline src/SPIN/fix_nve_spin.cpp]: This if statement is empty. Either something needs to be added inside the curly braces or the whole block can be removed.

File: src/SPIN/fix_nve_spin.cpp [hunk 2]
```diff
@@ -435,6 +451,7 @@ void FixNVESpin::ComputeInteractionsSpin(int i)
 
   double **sp = atom->sp;
   double **fm = atom->fm;
+  //double **fm_long = atom->fm_long;
```
Comment (akohlmey) [inline src/SPIN/fix_nve_spin.cpp]: please delete this comment

File: src/KSPACE/pppm_dipole_spin.cpp [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+{
+  if (me == 0) {
+    if (screen) fprintf(screen,"PPPMDipoleSpin initialization ...\n");
+    if (logfile) fprintf(logfile,"PPPMDipoleSpin initialization ...\n");
+  }
+
+  // error check
+
+  spinflag = atom->sp?1:0;
+  //qsum_qsq(0); // q[i] is probably not declared ?
```
Comment (akohlmey) [inline src/KSPACE/pppm_dipole_spin.cpp]: please remove commented out code.

File: src/KSPACE/pppm_dipole_spin.cpp [hunk 4]
```diff
@@ -0,0 +1,760 @@
...
+{
+  dipoleflag = 0;
+  spinflag = 1;
+  
+  hbar = force->hplanck/MY_2PI;         	// eV/(rad.THz)
+  mub = 9.274e-4;                     		// in A.Ang^2
+  mu_0 = 785.15;               			// in eV/Ang/A^2
+  mub2mu0 = mub * mub * mu_0 / (4.0*MY_PI);	// in eV.Ang^3
+  //mub2mu0 = mub * mub * mu_0 / (4.0*MY_PI);	// in eV
+  mub2mu0 = mub * mub * mu_0 / (4.0*MY_PI);	// in eV
```
Comment (akohlmey) [inline src/KSPACE/pppm_dipole_spin.cpp]: this is redundant

File: src/KSPACE/ewald_dipole_spin.cpp [hunk 5]
```diff
@@ -0,0 +1,858 @@
...
+  EwaldDipole(lmp)
+{
+  dipoleflag = 0;
+  spinflag = 1;
+  
+  hbar = force->hplanck/MY_2PI;         	// eV/(rad.THz)
+  mub = 9.274e-4;                     		// in A.Ang^2
+  mu_0 = 785.15;               			// in eV/Ang/A^2
+  mub2mu0 = mub * mub * mu_0 / (4.0*MY_PI);	// in eV.Ang^3
+  //mub2mu0 = mub * mub * mu_0 / (4.0*MY_PI);	// in eV
```
Comment (akohlmey) [inline src/KSPACE/ewald_dipole_spin.cpp]: please remove

**Comment by stanmoore1** [comment]
@akohlmey good feedback, we'll work on this. Thanks.

**Comment by stanmoore1** [comment]
Closes #70

**Comment by stanmoore1** [comment]
@akohlmey I addressed your feedback, this PR is ready for merge.

**Pull request merged**